ToolBox/CoinApi/CoinApiDataQueueHandler.cs

```diff
@@ -0,0 +1,360 @@
+
+                // CoinAPI requires at least 5 seconds between subscription requests so we need to batch them
+                _subscribedSymbols = symbolsToSubscribe.Union(_subscribedSymbols).ToHashSet();
+
+                ProcessSubscriptionRequest();
+            }
+        }
+
+        /// <summary>
+        /// Removes the specified symbols to the subscription
+        /// </summary>
+        /// <param name="job">Job we're processing.</param>
+        /// <param name="symbols">The symbols to be removed keyed by SecurityType</param>
+        public void Unsubscribe(LiveNodePacket job, IEnumerable<Symbol> symbols)
+        {
+            lock (_lockerSubscriptions)
+            {
+                var symbolsToUnsubscribe = (from symbol in symbols
+                                            where _subscribedSymbols.Contains(symbol)
+                                            select symbol).ToList();
+                if (symbolsToUnsubscribe.Count == 0)
+                    return;
+
+                Log.Trace($"CoinApiDataQueueHandler.Unsubscribe(): {string.Join(",", symbolsToUnsubscribe.Select(x => x.Value))}");
+
+                // CoinAPI requires at least 5 seconds between subscription requests so we need to batch them
+                _subscribedSymbols = _subscribedSymbols.Where(x => !symbolsToUnsubscribe.Contains(x)).ToHashSet();
+
+                ProcessSubscriptionRequest();
+            }
+        }
+
+        /// <summary>
+        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
+        /// </summary>
+        public void Dispose()
+        {
+            _connectionHandler.DisposeSafely();
+
+            if (_webSocket.IsOpen)
+            {
+                _webSocket.Close();
+            }
+        }
+
+        /// <summary>
+        /// Helper method used in QC backend
+        /// </summary>
+        /// <param name="markets">List of LEAN markets (exchanges) to subscribe</param>
+        public void SubscribeMarkets(List<string> markets)
+        {
+            if (markets.Count == 0) return;
+
+            Log.Trace($"CoinApiDataQueueHandler.SubscribeMarkets(): {string.Join(",", markets)}");
+
+            _subscribedExchanges = markets.ToList();
+
+            var message = JsonConvert.SerializeObject(new HelloMessage
+            {
+                ApiKey = _apiKey,
+                Heartbeat = true,
+                SubscribeDataType = new[] { "trade", "quote" },
+                SubscribeFilterSymbolId = markets.Select(x => _symbolMapper.GetExchangeId(x)).ToArray()
+            });
+
+            _webSocket.Send(message);
+
+            _connectionHandler.EnableMonitoring(true);
+        }
+
+        private void ProcessSubscriptionRequest()
+        {
+            if (_subscriptionsPending) return;
+
+            _lastSubscribeRequestUtcTime = DateTime.UtcNow;
+            _subscriptionsPending = true;
```

Comment:
At first this looks like a potential race condition, but this entire method is covered by `lock(_lockerSubsriptions` via the `Subscribe`/`Unsubscribe` methods.